quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -407,66 +302,171 @@ impl IngestController {
             successes: get_or_create_open_shards_successes,
             failures: get_or_create_open_shards_failures,
-        })
+        };
+        Ok(response)
+    }
+
+    /// Allocates and assigns new shards to ingesters.
+    fn allocate_shards(
+        &self,
+        num_shards_to_allocate: usize,
+        unavailable_leaders: &FnvHashSet<NodeId>,
+        model: &ControlPlaneModel,
+    ) -> Option<Vec<(NodeId, Option<NodeId>)>> {
+        let ingesters: Vec<NodeId> = self
+            .ingester_pool
+            .keys()
+            .into_iter()
+            .filter(|ingester| !unavailable_leaders.contains(ingester))
+            .sorted_by(|left, right| left.cmp(right))
+            .collect();
+
+        let num_ingesters = ingesters.len();
+
+        if num_ingesters == 0 {
+            warn!("failed to allocate {num_shards_to_allocate} shards: no ingesters available");
+            return None;
+        } else if self.replication_factor > num_ingesters {
+            warn!(
+                "failed to allocate {num_shards_to_allocate} shards: replication factor is \
+                 greater than the number of available ingesters"
+            );
+            return None;
+        }
+        let mut leader_follower_pairs = Vec::with_capacity(num_shards_to_allocate);
+
+        let mut num_open_shards: usize = 0;
+        let mut num_open_shards_per_ingester: HashMap<&str, usize> =
+            HashMap::with_capacity(num_ingesters);
+
+        for shard in model.all_shards() {
+            if shard.is_open() && !unavailable_leaders.contains(&shard.leader_id) {
+                num_open_shards += 1;
+
+                *num_open_shards_per_ingester
+                    .entry(&shard.leader_id)
+                    .or_default() += 1;
+            }
+        }
+        let mut num_remaining_shards_to_allocate = num_shards_to_allocate;
+        let num_open_shards_target = num_shards_to_allocate + num_open_shards;
+        let max_num_shards_to_allocate_per_node = num_open_shards_target / num_ingesters;
+
+        // Allocate at most `max_num_shards_to_allocate_per_node` shards to each ingester.
+        for (leader_id, follower_id) in ingesters.iter().zip(ingesters.iter().cycle().skip(1)) {
+            if num_remaining_shards_to_allocate == 0 {
+                break;
+            }
+            let num_open_shards_inner = num_open_shards_per_ingester
+                .get(leader_id.as_str())
+                .copied()
+                .unwrap_or_default();
+
+            let num_shards_to_allocate_inner = max_num_shards_to_allocate_per_node
+                .saturating_sub(num_open_shards_inner)
+                .min(num_remaining_shards_to_allocate);
+
+            for _ in 0..num_shards_to_allocate_inner {
+                num_remaining_shards_to_allocate -= 1;
+
+                let leader = leader_id.clone();
+                let mut follower_opt = None;
+
+                if self.replication_factor > 1 {
```

Comment:
I think the mirroring of the primary shards balanced allocation "shifted by one ingester" for replicas is broken because of the `unavailable_leaders` thing. I'll rework this part.